### DISCUSSION TITLE
How to get rid of Faults

### DISCUSSION THREAD

**Discussion by trehn** [bundlewrap/bundlewrap#690] [Ideas]
Faults for a long time have been both necessary and a pain to handle.

### Why do we need Faults?

* When looking at metadata for many nodes, retrieving the secrets in that metadata may be slow and not actually required.
* Not everyone might have access to all secrets, but they should still be able to apply items that don't need those secrets.
* Faults make it easy to scan metadata for secrets (for example to show all passwords for a node).
* Fault behavior can be inherited by custom solutions for secret storage (bundlewrap-teamvault).

### Why are Faults painful?

* We had to implement "id strings" to make them comparable without resolving them.
* Faults cannot be accurately represented when viewing metadata as JSON. (#557)
* Faults prevent metadata serialization and caching. (#564)
* Faults cannot be used in TOML nodes.

### Where might Faults need to be resolved?

This boils down to a list of places where `node.metadata` can be used:

**1) Reactors**

When returning partial metadata here, we don't need secrets resolved.

**2) items.py**

We may need *some* secrets here that we end up formatting into action commands. But insisting on resolving *all* secrets here would mean that users without access to all secrets won't even be able to generate a list of items (and thus lose the ability to even apply items they have access to).

**3) File templates and k8s manifests**

We need the secret here, but if we fail to get it, we know what item it's for and can just skip that item.

**4) CLI node lambda selectors**

Users would probably expect access to secrets here, but rarely actually use them.

**5) bw metadata**

Dumping metadata to JSON resolves Faults. This isn't always desirable because it makes it hard to find out where Faults are actually being used (they look like regular strings here).

### What can we do about it?

The solution most likely involves magic strings of some sort. Without magic strings, secret retrieval has to happen explicitly in every place a secret might be needed (i.e. inside a file template).

For a more general approach that requires no extra logic in templates, we need a way to retroactively insert secrets into metadata. This is where magic strings come in.

The difficulty with magic strings is finding the best time to resolve them. Faults have the benefit of only being resolved when they are actually used. With magic strings, this might be impossible.

Let's try and find a reasonable set of points where magic strings would be resolved for a node. We need to modify the logic behind `node.metadata` to *not* resolve magic strings in: reactors, node selectors, `bw metadata`. Accessing `node.metadata` from items.py or file templates *will* resolve *all* magic strings for a node. This means that it is no longer possible to apply a node when you don't have all the secrets. But at least you don't need secrets to every node. Might be a reasonable compromise.

**Comment by vain** [comment]
Some of the stuff we discussed today:

It should be safe to assume that Faults are only ever *resolved* inside of items (e.g., while rendering a file template, while executing an `action` command, while setting a postgres user password, …). The reason for this assumption is that this is the only time that a Fault makes sense (so an item can be skipped).

In other words, the true values of Faults are only *needed* inside of items. Or when manually inspecting them using commands like `bw metadata`.

However:

```
'/etc/foo': {
    'content_type': 'mako',
    'context': {
        'boom': node.metadata.get('my/fancy/fault'),
    },
},
```

Here, a Fault is being *referenced* in `items.py`. It would be wrong to *resolve* it in this call to `metadata.get()`, this must only be done later inside of the item.

So, we can’t simply do this: “Have all calls to `metadata.get()` resolve magic strings when called from inside an item.” (Because when used in `context` like above, there is no call to `metadata.get()` anymore.)

Also: How would `metadata.get()` know which strings are magic strings and which are not? Does the *entire* string have to match a certain pattern? Or is it legal if only a *substring* matches that pattern? (Example: Users might build strings like `$username:$password` which consist of *two* magic strings.)

It would be cleaner to not allow partial strings:

-   Legal: `metadata.get('foo') -> '!vault.password_for:foobar'`
-   Illegal: `metadata.get('foo') -> '!vault.decrypt:blauser:!vault.decrypt:blapassword'`

I think it would be *doable* to enforce this requirement, but I’m not sure if it’s *practical*.

If we *did* allow partial strings, though, then why not replace all magic strings in the *rendered* file item? This can be tricky due to collisions, of course.
